File: Olympiad/COCI/official/2007/contest1_solutions/slikar.cpp

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 1 - Task SLIKAR
  Programming language C++
*/

#include <cstdio>
#include <queue>
#include <cstring>
#include <utility>
using namespace std;

#define MAXN 50

int  r, s;
char karta[MAXN+1][MAXN+1];
int  poplava[MAXN+1][MAXN+1];
int  slikar[MAXN+1][MAXN+1];

int  potez[8][2] = { {1,0}, {-1,0}, {0,1}, {0,-1} };

queue<pair<int, int> > Q;

int main() {
   memset( poplava, -1, sizeof poplava );
   memset( slikar, -1, sizeof slikar );

   scanf( "%d%d", &r, &s);
   for ( int i=0; i<r; ++i )
      scanf( "%s", karta[i] );

   int sx=-1, sy=-1, dx=-1, dy=-1;

   // prolazak kroz kartu i nalazenje poplavljenih polja
   for ( int i=0; i<r; ++i )
      for ( int j=0; j<s; ++j ) {
         if ( karta[i][j] == 'S' ) {
            sx = i; sy = j;
            karta[i][j] = '.';
         }

         if ( karta[i][j] == 'D' ) {
            dx = i; dy = j;
         }

         if ( karta[i][j] == '*' ) poplava[i][j] = 0;
         if ( karta[i][j] == '.' ) {
            bool poplavljeno = false;
            for (int k = 0; k < 4; k++) {
               int ni = i + potez[k][0];
               int nj = j + potez[k][1];
               if ( ni < 0 || ni >= r ) continue;
               if ( nj < 0 || nj >= s ) continue;
               if ( karta[ni][nj] == '*' ) poplavljeno = true;
            }
            if ( poplavljeno ) {
               Q.push( make_pair(i, j) );
               poplava[i][j] = 1;
            }
         }

      }

   // poplava
   while ( !Q.empty() ) {
      int x = Q.front().first;
      int y = Q.front().second;
      Q.pop();

      for ( int k=0; k<4; ++k ) {
         int nx = x + potez[k][0];
         int ny = y + potez[k][1];
         if ( nx < 0 || nx >= r ) continue;
         if ( ny < 0 || ny >= s ) continue;
         if ( poplava[nx][ny] == -1 && karta[nx][ny] != 'X' && karta[nx][ny] != 'D' ) {
            poplava[nx][ny] = poplava[x][y] + 1;
            Q.push( make_pair( nx, ny ) );
         }
      }
   }

   // slikar
   slikar[sx][sy] = 0;
   Q.push( make_pair( sx, sy ) );

   while ( !Q.empty() ) {
      int x = Q.front().first;
      int y = Q.front().second;
      Q.pop();

      for ( int k = 0; k < 4; ++k ) {
         int nx = x + potez[k][0];
         int ny = y + potez[k][1];
         if ( nx < 0 || nx >= r ) continue;
         if ( ny < 0 || ny >= s ) continue;
         if ( slikar[nx][ny] == -1 && karta[nx][ny] != 'X' )
            if ( poplava[nx][ny] > slikar[x][y] + 1 || poplava[nx][ny] == -1 ) {
               slikar[nx][ny] = slikar[x][y] + 1;
               Q.push(make_pair(nx, ny));
            }
      }
   }

   if ( slikar[dx][dy] != -1 )
      printf( "%d\n", slikar[dx][dy] );
   else
      printf( "KAKTUS\n" );

   return 0;
}

File: Olympiad/COCI/official/2007/contest1_solutions/slikar.pas

(*
  Croatian Open Competition in Informatics 2006/2007
  Contest 1 - Task SLIKAR
  Programming language Pascal
*)

program slikar;

var
   i, j, r, s   : integer;
   stanje, novo : array[0..51, 0..51] of char;
   rj           : integer;
   gotovo       : boolean;

begin
   readln(r, s);
   for i:=1 to r do begin
      for j:=1 to s do
         read(stanje[i, j]);
      readln;
      stanje[i, 0]   := 'X';
      stanje[i, s+1] := 'X';
   end;

   for j:=0 to s+1 do begin
      stanje[0, j]   := 'X';
      stanje[r+1, j] := 'X';
   end;

   rj := 0;
   repeat
      rj := rj+1;

      novo := stanje;
      (* Oznaci polja na koja slikar moze doci *)
      for i:=1 to r do
         for j:=1 to s do
            if ( stanje[i, j-1] = 'S' ) or
               ( stanje[i, j+1] = 'S' ) or
               ( stanje[i-1, j] = 'S' ) or
               ( stanje[i+1, j] = 'S' ) then
            begin
               if stanje[i, j] = '.' then
                  novo[i, j] := 'S'
               else if stanje[i, j] = 'D' then begin
                  writeln( rj );
                  exit;
               end;
            end;
      stanje := novo;

      (* Prosiri poplavu *)
      for i:=1 to r do
         for j:=1 to s do
            if ( stanje[i, j-1] = '*' ) or
               ( stanje[i, j+1] = '*' ) or
               ( stanje[i-1, j] = '*' ) or
               ( stanje[i+1, j] = '*' ) then
            begin
               if stanje[i, j] in ['.', 'S'] then
                  novo[i, j] := '*';
            end;
      stanje := novo;

      (* Gotovi smo ako nema polja na kojem slikar moze biti (sve
         poplavljeno) *)
      gotovo := true;
      for i:=1 to r do
         for j:=1 to s do
            if stanje[i, j] = 'S' then
               gotovo := false;
   until gotovo;

   (* Pokoj im dusi *)
   writeln('KAKTUS');
end.
